README
NAME
    IO::Async::XMLStream::SAXReader - Dispatch SAX events from an XML
    stream.

VERSION
    version 0.001000

SYNOPSIS
        use IO::Async::XMLStream::SAXReader;
        use IO::Async::Loop;

        my $loop = IO::Async::Loop->new();

        my $sax  = IO::Async::XMLStream::SAXReader->new(
            handle => $SOME_IO_HANDLE,
            on_start_document => sub {
                my ( $saxreader, @args ) = @_;
                ...
            },
            on_start_element  => sub {
                my ( $saxreader, @args ) = @_;
                ...
            },
            on_end_document => sub {
                $loop->stop;
            },
        );

        $loop->add($sax);
        $loop->run();

    This sub-classes "IO::Async::Stream" to provide a streaming SAX parser.

    For the individual "SAX" events that can be listened for, see
    "XML::SAX::Base".

    All are prefixed with the "on_" prefix as constructor arguments.

    Alternatively, if you already have an "XML::SAX" handler class you wish
    to reuse:

        use IO::Async::XMLStream::SAXReader;
        use IO::Async::Loop;

        my $loop = IO::Async::Loop->new();

        my $sax  = IO::Async::XMLStream::SAXReader->new(
            handle => $SOME_IO_HANDLE,
            sax_handler => YourClass->new();
            on_read_eof => sub {
                $loop->stop;
            },
        );

        $loop->add($sax);
        $loop->run();

AUTHOR
    Kent Fredric <kentfredric@gmail.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by Kent Fredric
    <kentfredric@gmail.com>.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Release history for IO-Async-XMLStream-SAXReader

0.001003 2017-03-10T22:59:50Z 2ab930d

0.001002 2017-03-10T22:58:32Z 3b7e8e8
 - Bugfix: Revert overly aggressive Perl upgrade caused by a backfiring MinimumPerl test false-positiving in an .xml
   file
 - Purge cruft files

 [Dependencies::Stats]
 - Dependencies changed since 0.001001, see misc/*.deps* for details
 - develop: +1
 - runtime: ↓1
 - test: ↓1

0.001001 2017-03-10T21:07:34Z 1ead399
 - Bugfix: Updated tooling fixes -Ddefault_inc_excludes_dot test failures

 [Dependencies::Stats]
 - Dependencies changed since 0.001000, see misc/*.deps* for details
 - configure: +1 ↓1 (recommends: ↑1)
 - develop: +18 ↑4 -6 (recommends: +1, suggests: +2 -1)
 - runtime: ↑1 -1
 - test: +2 ↓1 -3 (recommends: ↑2 ↓1 -1)

0.001000 2014-03-25T15:17:46Z
 - First version.

 [Dependencies::Aggregate]
 - 5.010 : 27
 - 5.012 : 24
 - 5.019 : 12


Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.014.
use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.4205"
  },
  "configure_requires" => {
    "Module::Build" => "0.4205"
  },
  "dist_abstract" => "Dispatch SAX events from an XML stream.",
  "dist_author" => [
    "Kent Fredric <kentfredric\@gmail.com>"
  ],
  "dist_name" => "IO-Async-XMLStream-SAXReader",
  "dist_version" => "0.001000",
  "license" => "perl",
  "module_name" => "IO::Async::XMLStream::SAXReader",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "IO::Async::Stream" => 0,
    "IO::Async::XMLStream::SAXReader::DuckHandler" => 0,
    "XML::LibXML::SAX::ChunkParser" => "0.00007",
    "parent" => 0,
    "perl" => "5.008",
    "strict" => 0,
    "utf8" => 0,
    "warnings" => 0
  },
  "script_files" => [],
  "test_requires" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec::Functions" => 0,
    "IO::Async::Loop" => 0,
    "List::Util" => 0,
    "Test::More" => "1.001003",
    "version" => 0
  }
);


my %fallback_build_requires = (
  "ExtUtils::MakeMaker" => 0,
  "File::Spec::Functions" => 0,
  "IO::Async::Loop" => 0,
  "List::Util" => 0,
  "Module::Build" => "0.4205",
  "Test::More" => "1.001003",
  "version" => 0
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


maint/perlcritic.rc.gen.pl
#!/usr/bin/env perl
## no critic (Modules::RequireVersionVar)

# FILENAME: bundle_to_ini.pl
# CREATED: 02/06/14 01:48:56 by Kent Fredric (kentnl) <kentfredric@gmail.com>
# ABSTRACT: Write an INI file from a bundle

use 5.008;    #utf8
use strict;
use warnings;
use utf8;

use Carp qw( croak carp );
use Perl::Critic::ProfileCompiler::Util qw( create_bundle );

## no critic (ErrorHandling::RequireUseOfExceptions)
my $bundle = create_bundle('Example::Author::KENTNL');
$bundle->configure;

my @stopwords = (
  qw(
    )
);
for my $wordlist (@stopwords) {
  $bundle->add_or_append_policy_field( 'Documentation::PodSpelling' => ( 'stop_words' => $wordlist ) );
}

$bundle->remove_policy('ErrorHandling::RequireUseOfExceptions');

#$bundle->remove_policy('NamingConventions::Capitalization');

my $inf = $bundle->actionlist->get_inflated;

my $config = $inf->apply_config;

{
  open my $rcfile, '>', './perlcritic.rc' or croak 'Cant open perlcritic.rc';
  $rcfile->print( $config->as_ini, "\n" );
  close $rcfile or croak 'Something fubared closing perlcritic.rc';
}
my $deps = $inf->own_deps;
{
  open my $depsfile, '>', './perlcritic.deps' or croak 'Cant open perlcritic.deps';
  for my $key ( sort keys %{$deps} ) {
    $depsfile->printf( "%s~%s\n", $key, $deps->{$key} );
    *STDERR->printf( "%s => %s\n", $key, $deps->{$key} );
  }
  close $depsfile or carp 'Something fubared closing perlcritic.deps';
}



lib/IO/Async/XMLStream/SAXReader.pm
use 5.006;    # our
use strict;
use warnings;

package IO::Async::XMLStream::SAXReader;

our $VERSION = '0.001003';

# ABSTRACT: Dispatch SAX events from an XML stream.

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

use parent 'IO::Async::Stream';





















































use XML::LibXML::SAX::ChunkParser 0.00007;    # Buggy Finish
use IO::Async::XMLStream::SAXReader::DuckHandler;

## no critic (NamingConventions)
sub _SAXReader {
  my ($self) = @_;
  my $key = 'SAXReader';
  return $self->{$key} if exists $self->{$key};
  $self->{$key} = {};
  $self->{$key}->{Parser} = XML::LibXML::SAX::ChunkParser->new( Handler => $self->{sax_handler} );
  return $self->{$key};
}
## use critic

my @XML_METHODS = qw(
  attlist_decl
  attribute_decl
  characters
  comment
  doctype_decl
  element_decl
  end_cdata
  end_document
  end_dtd
  end_element
  end_entity
  end_prefix_mapping
  entity_decl
  entity_reference
  error
  external_entity_decl
  fatal_error
  ignorable_whitespace
  internal_entity_decl
  notation_decl
  processing_instruction
  resolve_entity
  set_document_locator
  skipped_entity
  start_cdata
  start_document
  start_dtd
  start_element
  start_entity
  start_prefix_mapping
  unparsed_entity_decl
  warning
  xml_decl
);

sub configure {
  my ( $self, %params ) = @_;

  for my $method ('sax_handler') {
    next unless exists $params{$method};
    $self->{$method} = delete $params{$method};
  }

  if ( not $self->{'sax_handler'} ) {
    $self->{'sax_handler'} = IO::Async::XMLStream::SAXReader::DuckHandler->new( { SAXReader => $self, }, );
    for my $method (@XML_METHODS) {
      next unless exists $params{ 'on_' . $method };
      $self->{ 'on_' . $method } = delete $params{ 'on_' . $method };
    }
  }
  $self->_SAXReader;
  return $self->SUPER::configure(%params);
}

sub on_read {
  my ( $self, $buffref, $eof ) = @_;
  my $text = substr ${$buffref}, 0, length ${$buffref}, q[];

  $self->_SAXReader->{Parser}->parse_chunk($text) if length $text;
  if ($eof) {
    $self->_SAXReader->{Parser}->finish;
    return 0;
  }
  return 1;
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

IO::Async::XMLStream::SAXReader - Dispatch SAX events from an XML stream.

=head1 VERSION

version 0.001003

=head1 SYNOPSIS

    use IO::Async::XMLStream::SAXReader;
    use IO::Async::Loop;

    my $loop = IO::Async::Loop->new();

    my $sax  = IO::Async::XMLStream::SAXReader->new(
        handle => $SOME_IO_HANDLE,
        on_start_document => sub {
            my ( $saxreader, @args ) = @_;
            ...
        },
        on_start_element  => sub {
            my ( $saxreader, @args ) = @_;
            ...
        },
        on_end_document => sub {
            $loop->stop;
        },
    );

    $loop->add($sax);
    $loop->run();

This sub-classes L<< C<IO::Async::Stream>|IO::Async::Stream >> to provide a streaming SAX parser.

For the individual C<SAX> events that can be listened for, see L<< C<XML::SAX::Base>|XML::SAX::Base >>.

All are prefixed with the C<on_> prefix as constructor arguments.

Alternatively, if you already have an L<< C<XML::SAX>|XML::SAX >> handler class you wish to reuse:

    use IO::Async::XMLStream::SAXReader;
    use IO::Async::Loop;

    my $loop = IO::Async::Loop->new();

    my $sax  = IO::Async::XMLStream::SAXReader->new(
        handle => $SOME_IO_HANDLE,
        sax_handler => YourClass->new();
        on_read_eof => sub {
            $loop->stop;
        },
    );

    $loop->add($sax);
    $loop->run();

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Kent Fredric <kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/IO/Async/XMLStream/SAXReader/DuckHandler.pm
use 5.006;    # our
use strict;
use warnings;

package IO::Async::XMLStream::SAXReader::DuckHandler;

our $VERSION = '0.001003';

# ABSTRACT: Deferred Handler proxy for IO::Async constructor-driven interface

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

use Scalar::Util qw(weaken);
use Carp qw(croak);










sub new {
  my ( $self, $opts ) = @_;
  croak('SAXReader option is mandatory') unless exists $opts->{SAXReader};
  weaken $opts->{SAXReader};
  return bless $opts, $self;
}

sub _dyn_method {
  my ( $self, $method ) = @_;
  my $sax      = $self->{SAXReader};
  my $event    = 'on_' . $method;
  my $callback = $sax->can_event($event);
  return unless $callback;
  return sub {
    my ( undef, @args ) = @_;
    return $callback->( $sax, @args );
  };
}

sub can {
  my ( $self, $method ) = @_;
  my $orig = $self->SUPER::can($method);
  return $orig if $orig;
  return $self->_dyn_method($method);
}

## no critic (ClassHierarchies::ProhibitAutoloading)
sub AUTOLOAD {
  my ( $self, @args ) = @_;
  ( my $methname = our $AUTOLOAD ) =~ s/.+:://msx;
  return if 'DESTROY' eq $methname;
  return unless my $meth = $self->_dyn_method($methname);
  return $meth->( $self, @args );
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

IO::Async::XMLStream::SAXReader::DuckHandler - Deferred Handler proxy for IO::Async constructor-driven interface

=head1 VERSION

version 0.001003

=for Pod::Coverage new
can

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Kent Fredric <kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.009.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Dispatch SAX events from an XML stream.",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "IO-Async-XMLStream-SAXReader",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "IO::Async::XMLStream::SAXReader",
  "PREREQ_PM" => {
    "Carp" => 0,
    "IO::Async::Stream" => 0,
    "Scalar::Util" => 0,
    "XML::LibXML::SAX::ChunkParser" => "0.00007",
    "parent" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "IO::Async::Loop" => 0,
    "Test::More" => 0
  },
  "VERSION" => "0.001003",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "IO::Async::Loop" => 0,
  "IO::Async::Stream" => 0,
  "Scalar::Util" => 0,
  "Test::More" => 0,
  "XML::LibXML::SAX::ChunkParser" => "0.00007",
  "parent" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
